draft-ietf-tls-esni.md

```diff
@@ -782,10 +777,30 @@ including servers which do not implement this specification.
 
 ### HelloRetryRequest
 
-It is an error for the client to offer ECH before the HelloRetryRequest but not
-after. Likewise, it is an error for the client to offer ECH after the
-HelloRetryRequest but not before. If either of these conditions occurs, then the
-client-facing server MUST abort the handshake with an "illegal_parameter" alert.
+In case a HelloRetryRequest (HRR) is sent, the client-facing server enforces the
+following invariant. Let CH1' denote the first ClientHello processed by the
+server and let CH2' denote the second: either CH1' == CH2' == ClientHelloOuter
+or CH1' == CH2' == ClientHelloInner. In particular, it is an error to process
+ClientHelloInner followed by ClientHelloOuter or vice versa, as this could
+trigger undefined protocol behavior.
+
+Let CH1 and CH2 denote, respectively, the first and second ClientHello
+transmitted on the wire by the client. (Note that these are not necessarily the
+same as the messages processed by the server.) To enforce this invariant, the
+client-facing server behaves as follows.
+
+1. If CH1 contains the "encrypted_client_hello" extension but CH2 does not, or
+   if CH2 contains the "encrypted_client_hello" extension but CH1 does not, then
+   the server MUST abort the handshake with an "illegal_parameter" alert.
+1. Suppose the "encrypted_client_hello" extension is sent in both CH1 and CH2,
+   each extension having a well-formed ClientECH structure as its payload. If
+   the configuration identifier (see {{ech-configuration}}) differs between CH1
+   and CH2, then the server MUST abort with an "illegal_parameter" alert.
```

Comment:
Does the server need to check the config identifier if it declines ECH, or just if it accepts ECH? What happens if the client and server use the trial decryption mode, where the config ID is empty and you only determine the config by using it? Perhaps this should be phrased in terms of the ECHConfig used. WDYT about:

> If the client-facing server accepts ECH in response to CH1, it MUST check that the ClientECH.config_id parameter is the same between CH1 and CH2. If they differ, the server MUST abort the connection with an "illegal_parameter" alert. It then MUST decrypt the extension using the previously-selected ECHConfig and, if decryption fails, abort the connection with a "decrypt_error" alert. Note that, if the connection used the trial decryption mode with an empty ClientECH.config_id value, the server does not continue to try other configurations on decryption error.

Though this means, if the server believes in trial decryption, you need to remember *both* the ECHConfig *and* whether to expect an empty or matching config_id value.

(As an aside, this admits another GREASE-distinguishing active attack, and one I don't see any way of avoiding. But it only applies to connections that would naturally trip HRR.)

Reply:
A semi-related change that might make it easier to phrase this paragraph: https://github.com/davidben/draft-ietf-tls-esni/commit/f9b3aaf56b9c759c3ba3ab3dee4f99da34b2e460

Reply:
> Does the server need to check the config identifier if it declines ECH, or just if it accepts ECH?

It needs to check the config in both cases.

> What happens if the client and server use the trial decryption mode, where the config ID is empty and you only determine the config by using it?

I hadn't considered this case. 

> A semi-related change that might make it easier to phrase this paragraph: https://github.com/davidben/draft-ietf-tls-esni/commit/f9b3aaf56b9c759c3ba3ab3dee4f99da34b2e460

I think this will help. Let's land that PR and then come back to this.